button-merchant/src/main/java/com/usebutton/merchant/SSLManagerImpl.java

```diff
@@ -0,0 +1,184 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+package com.usebutton.merchant;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.support.annotation.VisibleForTesting;
+
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.TrustManagerFactory;
+
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.security.KeyManagementException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateException;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Network helper class that provides {@link KeyStore}s and {@link SSLContext} for use in secure
+ * socket communications.
+ */
+class SSLManagerImpl implements SSLManager {
+
+    private static SSLManager instance;
+
+    private KeyStore keyStore;
+    private SSLContext sslContext;
+    private Exception pendingException;
+    private List<Certificate> certificates;
+
+    private final char[] password;
+    private final CertificateProvider provider;
+
+    static SSLManager getInstance(CertificateProvider provider) {
+        return getInstance(provider, null);
+    }
+
+    static SSLManager getInstance(CertificateProvider provider, char[] password) {
+        if (instance == null) {
+            instance = new SSLManagerImpl(provider, password);
+            return instance;
+        }
+
+        boolean hasDiffPass = !Arrays.equals(password, ((SSLManagerImpl) instance).getPassword());
+        boolean hasDiffProv = ((SSLManagerImpl) instance).provider != provider;
+        if (hasDiffProv || hasDiffPass) {
+            instance = new SSLManagerImpl(provider, password);
+        }
+
+        return instance;
+    }
+
+    @VisibleForTesting
+    SSLManagerImpl(@NonNull CertificateProvider provider, @Nullable char[] password) {
```

Comment:
No need to explicitly include `@NonNull` param annotations since this is the default according to `package-info.java`.